File: iOS/Pool/Sources/Pool/Creator/PoolFromLayoutCreatorViewModel.swift
import Foundation
import DataPool
import UI
import ViewingState

public class PoolFromLayoutCreatorViewModel : ObservableObject {
    let gamblerId: String

    let createPoolUseCase: CreatePoolUseCase

    @Published @MainActor private(set) var state: MutationState<CreatePoolModel> = .idle(emptyCreatePoolModel())

    public init(gamblerId: String, createPoolUseCase: CreatePoolUseCase) {
        self.gamblerId = gamblerId
        self.createPoolUseCase = createPoolUseCase
    }

    @MainActor
    func createPool(createPoolModel: CreatePoolModel) {
        Task {
            let currentCreatePoolModel = state.activeValue()
            state = .mutating(original: currentCreatePoolModel, updated: createPoolModel)

            let result = await createPoolUseCase.execute(createPoolInput: createPoolModel.toCreatePoolInput(ownerGamblerId: gamblerId))

            switch result {
            case .success(let createdPoolOutput):
                state = .mutated(
                    original: currentCreatePoolModel,
                    updated: createPoolModel.copy { builder in builder.poolId = createdPoolOutput.poolId }
                )
            case .failure(let error):
                state = .failure(original: currentCreatePoolModel, updated: createPoolModel, error: error)
            }
        }
    }

    @MainActor
    func reset() {
        state = .idle(state.activeValue())
    }
}

File: Android/pool/src/main/java/com/felipearpa/tyche/pool/creator/PoolFromLayoutCreatorViewModel.kt
package com.felipearpa.tyche.pool.creator

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.data.pool.application.CreatePool
import com.felipearpa.ui.state.MutationState
import com.felipearpa.ui.state.activeValue
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class PoolFromLayoutCreatorViewModel(
    private val createPool: CreatePool,
    private val gamblerId: String,
) : ViewModel() {
    private val _state = MutableStateFlow<MutationState<CreatePoolModel>>(
        MutationState.Idle(emptyCreatePoolModel()),
    )
    val state = _state.asStateFlow()

    fun createPool(createPoolModel: CreatePoolModel) {
        val currentCreatePoolModel = _state.value.activeValue()
        _state.value =
            MutationState.Mutating(original = currentCreatePoolModel, updated = createPoolModel)
        viewModelScope.launch {
            val result =
                createPool.execute(
                    createPoolInput = createPoolModel.toCreatePoolInput(
                        ownerGamblerId = gamblerId,
                    ),
                )
            result.onSuccess { createPoolOutput ->
                _state.value = MutationState.Mutated(
                    original = currentCreatePoolModel,
                    updated = createPoolModel.copy(poolId = createPoolOutput.poolId),
                )
            }.onFailure {
                _state.value = MutationState.Failure(
                    original = currentCreatePoolModel,
                    updated = createPoolModel,
                    exception = it,
                )
            }
        }
    }

    fun reset() {
        _state.update { MutationState.Idle(it.activeValue()) }
    }
}
